Fix for embedded PL/SQL blocks (Oracle Pro*C)

aeryomenko opened the pull request:
Cppcheck is able to isolate embedded SQL blocks written for Oracle Pro\*C precompiler with use of wrappers in the form of assembler inline sections. So any following constructions are skipped well:
`EXEC SQL ... ;`
However Pro\*C precompiler also allows to embed more complicated blocks containing unnamed procedures written in Oracle procedure language (PL/SQL) or even definitions of named Oracle objects, like procedures, functions and sequences. In both the cases embedded block syntax as follows [(docs.oracle.com)](https://docs.oracle.com/cd/B28359_01/appdev.111/b28427/pc_07pls.htm#i2332) :
```
EXEC SQL EXECUTE
DECLARE
   ... 
BEGIN 
   ... 
END; 
END-EXEC; 
```
Both DECLARE section (optional) and procedure body section can contain mutiple lines terminated with ';'.
Suggested patch allows to locate and isolate all of the embedded SQL or PL/SQL blocks.

amai2012 commented:
Is the code safe if it fails to detect the end?

aeryomenko commented:
Yes, it is safe. If the end is reached without finding any acceptable terminator (';' or END-EXEC keyword), null will be used as the right boundary for asm section.
Also the first assertion in the test case simplifySQL, which was written for previous implementation, proves this fact.

danmar commented:
if you fail to detect the end it seems to me it would be best that the simplification is not made. how would it work to skip the simplification if findSqlEnd returns null?

Example code:

    void f() {
        EXEC SQL 123 456
    }

I dont want that the `}` is included in the asm code.

aeryomenko commented:
I agree, '}' can't be used inside the SQL or PL/SQL block, so simplification can be stopped when this token is found. However this is of course a syntax error in the analyzed file.

danmar commented:
I still would prefer that your code is more safe and stops when anything unexpected is seen.

What should happen with this code:

    void f() {
        MACRO(EXEC SQL);
    }

Do we know that there is always a `;` `{` or `}` before the `EXEC`?

Are all other tokens except `}` allowed in the expression?

+= -= *= /= ... + - * / % & && | || ^ << >> ~ !

aeryomenko commented:
Regarding using of macro in Pro\*C code, there a lot of limitations, since this pre-compiler is called before the C preprocessor, so it has its own parser of preprocessor directives. At least using of macro is forbidden inside the EXEC SQL statements - [examples from specification](https://docs.oracle.com/database/121/LNPCC/pc_05adv.htm#LNPCC3293).
Inclusion of EXEC SQL directives or their parts in the macro body or macro arguments as in your example is not documented, but in practice it is impossible. So I don't think that it must be taken into account.
Concerning allowable tokens, yes some of them, like c-style assignments and comparison operators, can not be present in the EXEC SQL directive, but when cppcheck isolates the statement until them, some part of sebsequent expression will be lost. Maybe it would be better not to isolate such malformed directive, but to throw a syntax error instead?

aeryomenko commented:
I have added more operators to the exception list as was suggested earlier. Also I have changed behaviour in case when SQL block is not terminated properly - syntax error is thrown now.

danmar commented:
Sorry for the late reply!!

> Regarding using of macro in Pro*C code, there a lot of limitations

I am mostly concerned about the majority of users that don't use Pro*C. This simplification is enabled by default.

Imagine this plain-C program:

    #include <stdio.h>

     #define EXEC     "EXEC"
     #define SQL      "SQL"

     int main() {
        printf(EXEC SQL "...");
        return 0;
     }

As long as Cppcheck see those defines (if they are in the source file or in a header file that is found) then there is no problem.

I would like that you don't assume that "EXEC SQL" always means that the code is compiled with PRO*C. Try to sanity check this as much as possible and only simplify in proper PRO*C cases.

For information, we have had this simplification for a long time and I have not heard of problems. But I want to be careful anyway.

danmar commented:
I changed my mind. I think we will need to protect this simplification with a cfg file... so I will review your changes as they are now.. you don't need to change anything right now.

danmar commented:
I have updated the code. Cppcheck will by default not do anything about "EXEC SQL..". You will need to use `--library=pro_c_sql` on the command line from now on. It is also available in the GUI.

Do you think that the name `pro_c_sql` is good.. or should we rename it? `oracle_pl_sql` maybe?

aeryomenko commented:
Thank you for merging my change! It was really important for our projects, which use Oracle PL/SQL invocations.
Regarding the name of config file, I think it would be better to use something like `embedded_sql`, because as far as I saw there are few more SQL precompilers, which use the same basic syntax `EXEC SQL ... ;`. At least following DBMS have their own embedded SQL preprocessors: Oracle (Pro\*C), PostgreSQL (ECPG), HP NonStop SQL/MX. I used only Oracle Pro*C in my practice, but in Oracle documentation I had found mention that they support ANSI standard of embedded SQL (with their own extensions), but unfortunately I couldn't find this standard.
One more question about SQL precompilers, all of them use their own file extensions:
- \*.pc - [Oracle Pro\*C](https://docs.oracle.com/cd/A57673_01/DOC/api/doc/PC_22/ch03a.htm)
- \*.ec,\*.ecpp and few more - [SQL/MX](https://support.hpe.com/hpsc/doc/public/display?docId=emr_na-c02123996)
- \*.pgc - [PostgreSQL ECPG](https://www.postgresql.org/docs/9.1/static/ecpg-concept.html)

Currently we apply patch on top of release to include \*.pc files to the function `Path::isC()`. Is it possible to make some additional config file, where file extension can be matched with language?

danmar commented:
ok then it sounds good to rename it to embedded_sql.

> Currently we apply patch on top of release to include *.pc files to the function Path::isC(). Is it possible to make some additional config file, where file extension can be matched with language?

you can try this:

    cppcheck --language=c file1.pc

or put all files you want to check in filelist.txt and try:

    cppcheck --language=c --file-list=filelist.txt
